hathor/p2p/manager.py

```diff
@@ -107,6 +110,31 @@ def has_synced_peer(self) -> bool:
                 return True
         return False
 
+    def is_practically_synced(self) -> bool:
+        """ Returns whether we can consider for practical resons that we are synced to the network.
+
+        This method name is intentionally a bit vague to give room for some imprecisions and subjective protection
+        agains malicious or rougue nodes. This method is primarily intended for use by APIs that need to wait for the
+        sync to finish but can work when it hasn't (even with bad, but not catastrophic consequences).
+
+        For example: `get_block_template` API can work at anytime, but before the node has synced, it will lead to
+        blocks way into the past with too low difficulty (because of decay), which is bad, but not catastrophic.
+        """
+        # How it works: filter out the peers that seem to have a reasonable timestamp (not way into the future, and not
+        # way into the past). Test if we're synced with any of them.
+        timestamp = int(self.reactor.seconds())
+        # XXX: reusing settings parameter, should we have another param? does it really matter?
+        sync_threshold = settings.P2P_SYNC_THRESHOLD
+        reasonable_timestamp_range = range(timestamp - sync_threshold, timestamp + sync_threshold)
```

Comment:
This might give false negatives, I think. We update a peer's `synced_timestamp` with the latest block/tx we get from it. `settings.P2P_SYNC_THRESHOLD` is currently 1 minute, but we might have a period when it takes more than one minute to find a new block. In that case, even if a peer is synced, this function will return false.

Reply:
Yeah, I've tested and it actually happens quite often. I've made a specific param for this and set it to 600s (10m), which should be plenty and safe.

Reply:
We can have it as `n * AVG_TIME_BETWEEN_BLOCKS` and we can calculate the probability that we have n periods without any blocks.

Reply:
Another way to do this would be taking the highest timestamp that is not much into too in the future from now and using that as a reference. That way it doesn't matter if there are no blocks for 1h (in case of an extreme hashrate drop).

Reply:
Not sure I understand well your suggestion.

Reply:
Even using `n * AVG_TIME_BETWEEN_BLOCKS`, it may happen to find a long period without any blocks because the hash rate has plummet. Maybe we should use `MAX_DISTANCE_BETWEEN_BLOCKS`. What do you think?

Reply:
`MAX_DISTANCE_BETWEEN_BLOCKS` won't matter, there is an `n` that makes `n * AVG_TIME_BETWEEN_BLOCKS` value's higher than that, it could be better if the meaning of the constant made more sense to use here, but I think it doesn't

My proposal use the highest `synced_timestamp` (excluding nodes that are more than some `FUTURE_THRESHOLD` more than `datetime.now`), and use that as a reference. **Instead** of using `datetime.now` as a base for these cut-offs. Now `P2P_SYNC_THRESHOLD` could be reused, because it won't matter how long the network stays without blocks, because we're relative to whichever peer has highest timestamp. It's still possible for a peer to fake it's timestamp when the network does not have blocks, but we could eventually remove that peer from the `ready_connections` when it fails to provide a block with said timestamp, so these can still be independent.

What do you think? Does it make sense now? @msbrogli @obiyankenobi

Reply:
I had a wrong understanding of what `synced_timestamp` was. I've made an update to use `peer_timestamp`. Also done this to `is_synced`. These changes make much more sense to me know, but I still have to try them out.

Reply:
It makes more sense now.

I'm wondering how easy it is to make tests to check this heuristics. It'd be good to have them, as it's easy for us to come up with some different scenarios and the expected outcome. But I'm not sure about creating "fake" connections.

Reply:
I agree that it would be good to test the scenarios. We can just create the peers and fill the variables to generate the scenarios.